File: solutions/c/grains/1/grains.c
#include "grains.h"

uint64_t square(uint8_t index)
{
    if (index == 0 || index > 64)
        return (0);
    uint64_t result = 1;
    for (uint8_t i=1; i<index; i++)
        result *= 2;
    return (result);
}

uint64_t total(void)
{
    uint64_t result = 0;
    for (int i=1; i<=64; i++)
        result += square(i);
    return (result);
}

File: solutions/fsharp/grains/1/Grains.fs
module Grains

let square (n: int): Result<uint64,string> = 
    match n with
    | n when n > 64 -> Error "square must be between 1 and 64"
    | n when n < 1 -> Error "square must be between 1 and 64"
    | _ -> Ok (uint64 ((bigint 2) ** (n - 1)))
let total: Result<uint64,string> = Ok (uint64 (Seq.sumBy (fun x -> (bigint 2) ** (x - 1)) [1 .. 64]))